torch/jit.py

```diff
@@ -1,26 +1,375 @@
+        else:
+            return inner
+    if callable(arg):
+        return _compile(arg)
+    else:
+        return _compile
+
+
+def trace(arg=None, nderivs=0, params=tuple()):
+    """
+    Instrument a function or module for tracing, wrapping it in a
+    :class:`TracedModule`, whose forward accepts the same arguments as the
+    original function/module, but returns a tuple consisting of the
+    *trace* of an execution, as well as the original return value.
+
+    Tracing is guaranteed not to change the semantics of the function/module
+    that is traced.
+
+    Arguments:
+        arg (optional, torch.nn.Module or function): the function or module
+            to be traced.  If `None`, `trace` returns a decorator which can be
+            applied to the function or module you want to trace.
+        nderivs (int, default 0): the number of derivatives to trace.
+            Traces of derivatives are recorded into the same trace returned
+            after executing the `forward` of the resulting module, but
+            are not present until you run `backward()` (an appropriate
+            number of times) on the resulting model.
+        params (tuple of torch.nn.Parameter): extra parameters for a traced
+            function, which do not occur as arguments to the function in
+            question.  You generally do not need this for tracing modules, as
+            the parameters of a module are automatically computed.
+
+    Example: Trace as function decorator.
+
+        >>> @jit.trace
+        >>> def f(x);
+        >>>     return x * 2
+        >>> trace, out = f(Variable(torch.randn(1)))
+
+    Example: Trace as higher order function. (Notice that trace is a *curried*
+    function; you first apply it with the function/model to trace, and then
+    apply the result with the arguments.)
+
+        >>> traced_model = jit.trace(nn.LSTMCell())
+        >>> trace, out = traced_model(input, hidden)
+
+    Example: Trace the backwards pass as function decorator.
+
+        >>> @jit.trace(nderivs=1)
+        >>> def f(x);
+        >>>     return x * 2
+        >>> trace, out = f(Variable(torch.randn(1)))
+        >>> out.sum().backward()
+        >>> print(trace)
+
+    Example: Trace the backwards pass as higher order function.
+
+        >>> traced_model = jit.trace(nn.LSTMCell(), nderivs=1)
+        >>> trace, out = traced_model(input, hidden)
+        >>> out.sum().backward()
+        >>> print(trace)
+
+    Example: Trace a function with extra parameters. (If you trace
+    a Module, parameters are automatically computed via `state_dict`).
+
+        >>> lstm = nn.LSTMCell(10, 20)
+        >>> @jit.trace(params=lstm.parameters())
+        >>> def f(a, b):
+        >>>     return lstm(a, b)
+    """
+    # TODO: handle decorating a class (not a callable)
+    def _trace(inner):
+        return TracedModule(inner, nderivs=nderivs, params=params)
```

Comment:
This seems very easy to break, because classes are callable. This decorator should work differently for functions/module instances and module classes. You can't do this right now:
```python
@torch.jit.compile
class MyModule(Module):
    pass
```

Reply:
Yeah, I'm planning to test for classes and support this usage.